### PR TITLE
[Modules] Converted Network NetworkWatcher to new dependencies approach

### PR DESCRIPTION
**Pull request by AlexanderSehr**
# Description

- Converted Network NetworkWatcher to new dependencies approach
- Update removal logic to NOT remove the Resource Group `NetworkWatcherRG` as otherwise the tests would conflict

| Pipeline |
| - |
| [Network: NetworkWatchers](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.networkwatchers.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: utilities/pipelines/resourceRemoval/helper/Remove-Deployment.ps1 [hunk 1]
```diff
@@ -117,12 +119,16 @@ function Remove-Deployment {
         # ===============================
         $dependencyResourceNames = Get-DependencyResourceNameList
 
-        if ($resourcesToIgnore = $resourcesToRemove | Where-Object { (Split-Path $_.resourceId -Leaf) -in $dependencyResourceNames }) {
+        $resourceIdsToIgnore = @(
```
Comment (eriqua) [inline utilities/pipelines/resourceRemoval/helper/Remove-Deployment.ps1:122]: I'd need to go through the change a bit more in depth, but for now I'd suggest to add a paragraph in the [resource removal](https://github.com/Azure/ResourceModules/wiki/The%20CI%20environment%20-%20Deployment%20validation#resource-removal) Wiki section, mentioning exceptions. Currently it's only the `NetworkWatcherRG` but I'm planning to also update that section with some of the resources part of the azure security center module

Comment (AlexanderSehr) [reply utilities/pipelines/resourceRemoval/helper/Remove-Deployment.ps1:122]: Absolutly agree, great idea. I think this section (both in the Wiki & code) will anyways be updated to quite an extend once all dependencies converted and can be removed

Comment (AlexanderSehr) [reply utilities/pipelines/resourceRemoval/helper/Remove-Deployment.ps1:122]: Updated :)

File: modules/Microsoft.Network/networkWatchers/.test/min/deploy.test.bicep [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// ============== //
+// Test Execution //
+// ============== //
+#disable-next-line no-hardcoded-location // Disabled as the default RG & location are created in always one location, but each test has to deploy into a different one
+var testLocation = 'northeurope'
+module testDeployment '../../deploy.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name)}-test-${serviceShort}'
+  params: {
+    location: testLocation
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/networkWatchers/.test/min/deploy.test.bicep:36]: Add comment: that this is actually not required and only here for testing purposes

Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/min/deploy.test.bicep:37]: addressed

File: modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            osDisk: {
+                deleteOption: 'Delete'
+                createOption: 'FromImage'
+            }
+        }
+        hardwareProfile: {
+            vmSize: 'Standard_B1ms'
+        }
+        osProfile: {
+            adminUsername: '${virtualMachineName}cake'
```
Comment (eriqua) [inline modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep:106]: Hey.

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    ]
+    flowLogs: [
+      {
+        enabled: false
+        storageId: diagnosticDependencies.outputs.storageAccountResourceId
+        targetResourceId: resourceGroupResources.outputs.firstNetworkSecurityGroupResourceId
+      }
+      {
+        formatVersion: 1
+        name: 'adp-<<namePrefix>>-nsg-x-apgw-flowlog'
```
Comment (eriqua) [inline modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep:126]: Please update with new convention

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep [hunk 5]
```diff
@@ -0,0 +1,144 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+@description('Optional. The name of the resource group to deploy for testing purposes')
```
Comment (eriqua) [inline modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep:6]: Please add description closing dots

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+var testLocation = 'westeurope'
+module testDeployment '../../deploy.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name)}-test-${serviceShort}'
+  params: {
+    name: 'NetworkWatcher_${testLocation}'
+    location: testLocation
+    connectionMonitors: [
+      {
+        name: 'adp-<<namePrefix>>-conmon-${serviceShort}-x-001'
```
[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep:67]: done

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep (hunk 5 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: done

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep (hunk 4 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: done

File: modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep (hunk 3 above)
Comment (eriqua) [reply modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep:106]: By self reference? Always

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep (hunk 4 above)
Comment (eriqua) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: Is this a dependency or a sub resource? If a dependency, shouldn't it be referenced from the dependencies module output?

File: modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep (hunk 3 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep:106]: not anymore :P Besides, the computername wasn't MrM... I think

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep (hunk 4 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: I just kept the name of the original test :D I guess it's not a dependency then, which makes the name a bit odd. Will double check

Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: So there is no such dependency. Renamed it now to something non-dependency looking and updated the readme

File: modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep (hunk 3 above)
Comment (eriqua) [reply modules/Microsoft.Network/networkWatchers/.test/common/dependencies.bicep:106]: So it was a lie

File: modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep (hunk 4 above)
Comment (eriqua) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: Those look like child resource names. I'd suggest <<namePrefix>>-${short}-fl-001 and <<namePrefix>>-${short}-cm-001. Not sure why appgw is mentioned

Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: True, they are the children - deployed via child modules.

Comment (AlexanderSehr) [reply modules/Microsoft.Network/networkWatchers/.test/common/deploy.test.bicep]: Updated the names

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**